### PR TITLE
Add warning for combined `merge_to_*` turn lanes

### PR DESCRIPTION
**Pull request by Famlam**
Adds a warning for combined `;`-separated lanes in turn:lanes where one of the options is `merge_to_*`

For example: `turn:lanes=none|none|merge_to_left;slight_right`
https://taginfo.openstreetmap.org/tags/turn%3Alanes=none%7Cnone%7Cmerge_to_left%3Bslight_right#overview

I checked a dozen of them, none of them reflect the actual turn lanes; nearly all are highway exits with another exit upcoming, but no combined arrows on the highway telling people to merge_to_left or turn to right. All of them either have no arrows or the arrows start after the merge_to_left point. About 1000 cases worldwide.

If you feel uncomfortable with this one, feel completely free to reject this pull request; it's just something I noticed yesterday while going through the code and thought I could propose to warn about. I can understand why people would add this (especially on highway entrances with another exit upcoming), but since `turn:lanes` is representing the *indicated* turn lanes and there's no "merge_or_turn" symbol on ways, I think this'll always be an incorrect value of a `turn:lanes` lane...

### CODE REVIEW

File: plugins/Highway_Lanes.py [hunk 1]
```diff
@@ -107,6 +111,9 @@ def way(self, data, tags, nds):
                                 err.append({"class": 31606, "subclass": 0 + stablehash64(tl + '|' + t + '|' + str(i)), "text": T_("Unknown turn lanes value \"{0}\"", t)})
                             if (t == "merge_to_left" and i == 0) or (t == "merge_to_right" and i == len(ttt) - 1):
                                 err.append({"class": 31600, "subclass": 1 + stablehash64(tl + '|' + t + '|' + str(i))})
+                            elif (t != tt and t[0:9] == "merge_to_"):
```
Comment (frodrigo) [inline plugins/Highway_Lanes.py:114]: Could a false positive with multiple `"merge_to_` ?

Comment (frodrigo) [reply plugins/Highway_Lanes.py:114]: Also, check the other is a turn lane.

Comment (Famlam) [reply plugins/Highway_Lanes.py:114]: Thanks for the review!

> Could a false positive with multiple "merge_to_ ?

You mean a single lane to have `merge_to_right` and `merge_to_left`? I.e. `left|merge_to_right;merge_to_left|right`? (I think duplicate values on a single lane are invalid, right? Like `through|merge_to_left;merge_to_left`)

> Also, check the other is a turn lane.

I don't understand what you mean, sorry. 
If I'm not mistaken, the only other options of `t` here in this part of the loop would be [slight_/sharp_]right/left, through, reverse (or none, but that's invalid here I think, as `none` implies there's no lane marking).

Comment (Famlam) [reply plugins/Highway_Lanes.py:114]: > I don't understand what you mean, sorry.

Or do you mean I should check vs. `unknown` being False?

Comment (frodrigo) [reply plugins/Highway_Lanes.py]: Ok I reread the code. I was confused by the check into the loop.

You can do it before the loop. Do the split before the loop and check in the list if there is a `merge_to_` and other values.

Comment (Famlam) [reply plugins/Highway_Lanes.py]: Ah I just read this comment too late and went for another approach with (I think) the same result; 
I don't know if you'd like me to change it?

I think your approach would be (dummy code):
- take `set(tt.split(";"))` out of the loop (let's call it X here)
- `if ((len(X) > 2 and ("merge_to_right" in X or "merge_to_left" in X)) or ("merge_to_right" in X and not "merge_to_left" in X) or (not "merge_to_right" in X and "merge_to_left" in X)):` -> error

(If so, I'll also move the other merge_to_* check out of the loop)

Comment (frodrigo) [reply plugins/Highway_Lanes.py]: Yes, please, move out of the loop.

Comment (Famlam) [reply plugins/Highway_Lanes.py]: 👍 Done!
Better name suggestions for settt ("set of tt") are welcome ;)

Comment (Famlam) [reply plugins/Highway_Lanes.py]: The `t` in the `subclass` calculation must be replaced by something else now, as `t` is a loop variable, but I'm not entirely sure what to do with it. Can I just put `tt` there, or could that result in non-unique subclasses?

Comment (Famlam) [reply plugins/Highway_Lanes.py]: I think, in the vast majority of cases, t == tt, so the effect should be small

File: plugins/Highway_Lanes.py [hunk 2]
```diff
@@ -101,12 +105,23 @@ def way(self, data, tags, nds):
...
+
+                        if ("merge_to_left" in settt and i == 0) or ("merge_to_right" in settt and i == len(ttt) - 1):
+                            # A lane must exist in the merging direction
+                            err.append({"class": 31600, "subclass": 1 + stablehash64(tl + '|' + tt + '|' + str(i))})
+
+                        elif (not unknown and len(settt) > 1 and
+                              ((len(settt) > 2 and ("merge_to_right" in settt or "merge_to_left" in settt)) or
+                               ("merge_to_right" in settt and not "merge_to_left" in settt) or
+                               ("merge_to_left" in settt and not "merge_to_right" in settt))):
+                            # a combination of merge_to_* with a turn (other than another merge_to_*)
```
Comment (frodrigo) [inline plugins/Highway_Lanes.py:122]: Need to check other value are turn. It could be `through`, `none`...

Comment (Famlam) [reply plugins/Highway_Lanes.py:122]: I'm sorry, but I don't understand what you mean. It's checked vs `not unknown`, isn't that what you mean?

Comment (Famlam) [reply plugins/Highway_Lanes.py:122]: I've renamed the variable `unknown` in the hope that that makes things more clear?

Comment (frodrigo) [reply plugins/Highway_Lanes.py:122]: `through`, `none` are known value, but there are no turn.

Comment (Famlam) [reply plugins/Highway_Lanes.py:122]: But `merge_to_right;through|through` would be an non-existing combination, correct? Just like `merge_to_right;left|right` or anything else. That's just a "through" or a "left" lane, respectively?

And `merge_to_right;none|through` would even be invalid (how can a lane be both "none" and "something" (where something in this example is `merge_to_right`). (Same for `merge_to_right;|through`)

Comment (Famlam) [reply plugins/Highway_Lanes.py]: I added an explicit check versus a lane containing `none;anything`, where `anything` could be one of `left/slight_left/sharp_left/through/reverse/right/sharp_right/slight_right/merge_to_left/merge_to_right`

**Comment by frodrigo** [comment]
OK. I run it over taxes. Plase can you results:

316011 raises ways id:
```
<way id="90232728" version="12" user="">
<way id="442014567" version="6" user="">
<way id="476773038" version="4" user="">
<way id="476773039" version="4" user="">
<way id="566921602" version="8" user="">
<way id="784116755" version="2" user="">
```

316012 raises ways id:
```
<way id="9940745" version="7" user="">
<way id="27101115" version="7" user="">
<way id="28470249" version="6" user="">
<way id="53282516" version="8" user="">
<way id="427605390" version="6" user="">
<way id="505332678" version="11" user="">
<way id="539610977" version="4" user="">
<way id="602247668" version="3" user="">
<way id="603215109" version="2" user="">
<way id="603518024" version="2" user="">
<way id="603518142" version="2" user="">
<way id="782543697" version="2" user="">
<way id="892664882" version="1" user="">
<way id="893134243" version="3" user="">
<way id="913094495" version="1" user="">
```

**Comment by Famlam** [comment]
> OK. I run it over taxes. Plase can you results:

I've checked all the results. It seems all OK to me

Summary:
316011: no false positives, in all cases the `turn:lanes` do not reflect the actual signs/road markings
316012: no false positives; most of them were fixed in the past 2 weeks already (bad luck... but trust me, it wasn't me ;) ) but I guess that means these were also valid warnings as I assume those users wouldn't have changed the turn:lanes.

=== 316011 ===
`<way id="90232728" version="12" user="">`
has `turn:lanes=|||merge_to_left;slight_right`
[Mapillary](https://www.mapillary.com/app/?lat=32.926357762792&lng=-96.820739015194&z=17&dateFrom=2018-01-01&pKey=840063383591137&focus=photo&x=0.34674193715446433&y=0.5874410894442882&zoom=0) shows that the merge markings only starts past the exit

`<way id="442014567" version="6" user="">`
has `turn:lanes=slight_left;merge_to_right||`
[Mapillary](https://www.mapillary.com/app/?lat=29.89233800000001&lng=-95.41172899999998&z=16.16616695923874&pKey=823682468533126&focus=photo&dateFrom=2018-01-01&x=0.38172816986184627&y=0.6588906754458214&zoom=2.085686444230395) shows no signs/markings/... indicating a merge. Ahead, the lane turns into a mandatory left, before that, a vehicle can switch lanes at will without traffic sign

`<way id="476773038" version="4" user="">`
has `turn:lanes=left;merge_to_right|`
[Mapillary](https://www.mapillary.com/app/?lat=29.832705&lng=-95.37609&z=17&pKey=920039728563997&focus=photo&dateFrom=2018-01-01&x=0.5680183521085821&y=0.7649123527162324&zoom=0.5271401551031818) shows traffic signs ahead that would be `left|through;right` and nothing on the section before that

`<way id="476773039" version="4" user="">`
has `turn:lanes=left;merge_to_right|`
[Mapillary](https://www.mapillary.com/app/?lat=29.831472&lng=-95.375979&z=17&pKey=6117810868232629&focus=photo&dateFrom=2018-01-01&x=0.531622493146466&y=0.7221175973943957&zoom=0.33170316847613573) shows traffic signs ahead that would be `left|through;right` and `left only` on the lane in the section before that.

`<way id="566921602" version="8" user="">`
has `turn:lanes=||merge_to_left;right`
[Mapillary](https://www.mapillary.com/app/?lat=32.871829&lng=-96.765192&z=17&pKey=302330644837625&focus=photo&dateFrom=2018-01-01&x=0.5113503080324028&y=0.6133003266563188&zoom=0) shows only a pre-warning that a merge will occur *ahead*, but the actual merge sign starts at the turn.

`<way id="784116755" version="2" user="">`
has `turn:lanes=slight_left;merge_to_right|||`
[Mapillary](https://www.mapillary.com/app/?lat=29.744655&lng=-95.558372&z=17&pKey=795796554657533&focus=photo&dateFrom=2018-01-01&x=0.3764653581395958&y=0.6186246828404638&zoom=1.5247410465850655) shows merging to right is disallowed per traffic signs, left lane = only left


=== 316012 ===
`<way id="9940745" version="7" user="">`
Was fixed in https://www.openstreetmap.org/changeset/108858797 this week

`<way id="27101115" version="7" user="">`
Was fixed today in https://www.openstreetmap.org/changeset/108962044

`<way id="28470249" version="6" user="">`
Was fixed in https://www.openstreetmap.org/changeset/108858607 this week

`<way id="53282516" version="8" user="">`
Was fixed 9 days ago https://www.openstreetmap.org/changeset/108480375

`<way id="427605390" version="6" user="">`
Was fixed two months ago https://www.openstreetmap.org/changeset/105968585

`<way id="505332678" version="11" user="">`
Was fixed 10 days ago in https://www.openstreetmap.org/changeset/108400465

`<way id="539610977" version="4" user="">`
Was fixed this week in https://www.openstreetmap.org/changeset/108855809

`<way id="602247668" version="3" user="">`
Was fixed 10 days ago in https://www.openstreetmap.org/changeset/108400465

`<way id="603215109" version="2" user="">`
has `turn:lanes=left;through|through|through|through|right;||||through;right` (note the `;` in `right;|||`)
[Mapillary](https://www.mapillary.com/app/?lat=32.757084&lng=-97.334924&z=17&dateFrom=2018-01-01&pKey=523530945330052&focus=photo&x=0.44077116618985457&y=0.47443677040884236&zoom=0.7623705232925327) shows 3 lanes without marking (definitely not 9 lanes).

`<way id="603518024" version="2" user="">`
Was fixed 10 days ago in https://www.openstreetmap.org/changeset/108400465

`<way id="603518142" version="2" user="">`
has `turn:lanes=left;through|through|through|through|right;||||through;right`
[Mapillary](https://www.mapillary.com/app/?lat=32.757548470675&lng=-97.333878545583&z=17&dateFrom=2018-01-01&pKey=336597761146276&focus=photo&x=0.49308925984336227&y=0.5604573553652789&zoom=0) shows `left|left;through|none|none` (and only 4 lanes)

`<way id="782543697" version="2" user="">`
Was fixed 10 days ago in https://www.openstreetmap.org/changeset/108400465

`<way id="892664882" version="1" user="">`
Was fixed 10 days ago in https://www.openstreetmap.org/changeset/108398846

`<way id="893134243" version="3" user="">`
Was fixed 11 days ago in https://www.openstreetmap.org/changeset/108388357

`<way id="913094495" version="1" user="">`
Was fixed 12 days ago in https://www.openstreetmap.org/changeset/108344356

**Comment by frodrigo** [comment]
OK. Thank you.

**Comment by frodrigo** [comment]
Merged. But not deployed yet.

**Comment by Famlam** [comment]
Great! Thanks!

**Comment by jocelynj** [comment]
It was just deployed on osm-fr backends. Results should be updated in the next days.

**Comment by frodrigo** [comment]
All result here
http://osmose.openstreetmap.fr/en/issues/open?item=3160&class=316011,316012

**Pull request merged**